app/src/main/java/com/sdpteam/connectout/post/model/PostFirebaseDataSource.java

```diff
@@ -1,63 +1,169 @@
-                    for (DataSnapshot child : t.getResult().getChildren()) {
-                        Post post = child.getValue(Post.class);
-                        posts.add(post);
-                    }
-                    future.complete(new Result<>(posts, true, ""));
-                });
+        return eventIdsUserCanAccess(userId).thenCompose(allEventsUserCanAccess -> {
+            final CompletableFuture<Result<List<Post>>> result = new CompletableFuture<>();
 
-        return future;
+            FirebaseDatabase.getInstance().getReference().child(POSTS).get().addOnCompleteListener(task -> {
+                if (task.isSuccessful() && task.getResult().exists() && !forceFail) {
+                    final DataSnapshot dataSnapshot = task.getResult();
+                    final List<Post> postList = StreamSupport.stream(dataSnapshot.getChildren().spliterator(), false)
+                            .map(childSnapshot -> childSnapshot.getValue(Post.class))
+                            .filter(Objects::nonNull)
+                            .filter(post -> {
+                                if (post.getVisibility() != null && post.getVisibility().equals(SEMIPRIVATE) && post.getEventId() != null) {
+                                    return allEventsUserCanAccess.contains(post.getEventId());
+                                } else {
+                                    return true;
+                                }
+                            })
+                            .collect(Collectors.toList());
+                    result.complete(new Result<>(postList, true));
+                } else {
+                    result.complete(new Result<>(null, false, "Post fetch failed"));
+                }
+            });
+            return result;
+        });
     }
-    //TODO fix through firebase task
+
+    private CompletableFuture<List<String>> eventIdsUserCanAccess(String userId) {
+        final EventParticipantIdFilter organizerOrParticipant = new EventParticipantIdFilter(userId);
+
+        return new EventFirebaseDataSource()
+                .getEventsByFilter(organizerOrParticipant, ProfilesFilter.NONE)
+                .thenApply(events -> events.stream().map(Event::getId).collect(Collectors.toList()));
+    }
+
     @Override
     public CompletableFuture<Result<List<Post>>> fetchAllPostsOfEvent(String userId, String eventId) {
-        return fetchAllPosts(userId);
+        return fetchAllPostsFiltered(userId, post -> post.getEventId().equals(eventId));
     }
 
-    //TODO fix through firebase task
     @Override
     public CompletableFuture<Result<List<Post>>> fetchPostMadeByUser(String userId, String authorId) {
-        return fetchAllPosts(userId);
+        return fetchAllPostsFiltered(userId, post -> post.getProfileId().equals(authorId));
+    }
+
+    private CompletableFuture<Result<List<Post>>> fetchAllPostsFiltered(String userId, Predicate<Post> postPredicateFilter) {
```

Comment:
I would rename postPredicateFilter to postFilter simply

Reply:
Yeah, was struggling finding a clear name ahah